marimo/_save/save.py

```diff
@@ -37,35 +45,309 @@
+        self.scoped_refs |= set(f_locals.keys()) - set(ctx.globals.keys())
+        # Defined in the cell, and currently available in scope
+        self.scoped_refs |= ctx.graph.cells[cell_id].defs & set(
+            ctx.globals.keys()
+        )
+        # The defined private variables of this cell, normalized
+        self.scoped_refs |= set(
+            unmangle_local(x).name
+            for x in ctx.globals.keys()
+            if is_mangled_local(x, cell_id)
+        )
+
+        graph = ctx.graph
+        cell_id = ctx.cell_id or ctx.execution_context.cell_id
+        module = strip_function(self.fn)
+
+        self.base_block = BlockHasher(
+            module=module,
+            graph=graph,
+            cell_id=cell_id,
+            scope=self.scope,
+            pin_modules=self.pin_modules,
+            hash_type=self.hash_type,
+            scoped_refs=self.scoped_refs,
+            apply_content_hash=False,
+        )
+
+        # Load global cache from state
+        name = self.fn.__name__
+        # Note, that if the function name shadows a global variable, the
+        # lifetime of the cache will be tied to the global variable.
+        # We can invalidate that by making an invalid namespace.
+        if ctx.globals != f_locals:
+            name = name + "*"
+
+        context = "cache"
+        self._loader = ctx.state_registry.lookup(name, context=context)
+        if self._loader is None:
+            loader = MemoryLoader(name, max_size=self.max_size)
+            self._loader = State(loader, _name=name, _context=context)
+        else:
+            self._loader().resize(self.max_size)
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+        # Capture the deferred call case
+        if self.fn is None:
+            if len(args) != 1:
+                raise TypeError(
+                    "cache() takes at most 1 argument (expecting function)"
+                )
+            self._set_context(args[0])
+            return self
+
+        # Capture the call case
+        arg_dict = {k: v for (k, v) in zip(self._args, args)}
+        scope = {**self.scope, **get_context().globals, **arg_dict, **kwargs}
+        assert self._loader is not None, UNEXPECTED_FAILURE_BOILERPLATE
+        attempt = content_cache_attempt_from_base(
+            self.base_block,
+            scope,
+            self._loader(),
+            scoped_refs=self.scoped_refs,
+            as_fn=True,
+        )
+
+        if attempt.hit:
+            attempt.restore(scope)
+            return attempt.meta["return"]
+        response = self.fn(*args, **kwargs)
+        # stateful variables may be global
+        scope = {k: v for k, v in scope.items() if k in attempt.stateful_refs}
+        attempt.update(scope, meta={"return": response})
+        self._loader().save_cache(attempt)
+        return response
+
+
+def cache(
+    _fn: Optional[Callable[..., Any]] = None,
+    *,
+    pin_modules: bool = False,
+    hash_type: str = "sha256",
```

Comment:
@dmadisetti okay if we omit `hash_type` from the public API for now? or is there a compelling reason to include?

Reply:
Yep. I was checking for speedup against md5 etc. no noticable difference in my basic testing though, I don't think there's a meaningfully different choice